### ISSUE TITLE
Array.length type guard for array spreading

### ISSUE THREAD

**Issue by MidnightDesign** [Microsoft/TypeScript#28837]
**TypeScript Version:**  3.3.0-dev.20181204

**Search Terms:** array length type guard spread

**Code**

```ts
const timeString = '09:00';
const date = new Date();
const timeParts = timeString.split(':').map((part) => parseInt(part, 10));
if (timeParts.length > 0) {
    date.setHours(...timeParts);
}
```

**Expected behavior:** No errors

**Actual behavior:** TS2556: Expected 1-4 arguments, but got 0 or more.

**Playground Link:** https://www.typescriptlang.org/play/#src=const%20timeString%20%3D%20'09%3A00'%3B%0D%0Aconst%20date%20%3D%20new%20Date()%3B%0D%0Aconst%20timeParts%20%3D%20timeString.split('%3A').map((part)%20%3D%3E%20parseInt(part%2C%2010))%3B%0D%0Aif%20(timeParts.length%20%3E%200)%20%7B%0D%0A%20%20%20%20date.setHours(...timeParts)%3B%0D%0A%7D%0D%0A

**Related Issues:** None

**Comment by MartinJohns** [comment]
You could write a helper type-guard to make sure your array is a matching tuple, for example:

    function isTuple<T>(arr: T[], minLength: 1): arr is [T];
    function isTuple<T>(arr: T[], minLength: 2): arr is [T, T];
    function isTuple<T>(arr: T[], minLength: 3): arr is [T, T, T];
    function isTuple<T>(arr: T[], minLength: 4): arr is [T, T, T, T];
    function isTuple<T>(arr: T[], minLength: number) {
        return arr.length >= minLength;
    }
    
    if (isTuple(timeParts, 1)) {
        date.setHours(...timeParts);
    }

**Comment by MidnightDesign** [comment]
@weswigham Is that "Awaiting More Feedback" directed towards me? Of course there are a couple of workarounds. It would just be nice for the language/type system to be smart enough by itself.

**Comment by bopandrade** [comment]
Same issue here:

```ts
function looper(arg1: string, ...args:string[]) {
  console.log(arg1);
  if (args.length) return;
  looper(...args);
}
```

```
(parameter) args: string[]
Expected at least 1 arguments, but got 0 or more.
```

Although in my example there is a simple workaround since i only use one type:
```ts
function looperWorkaround(arg1: string, arg2?: string, ...args:string[]) {
  console.log(arg1);
  if (!arg2) return;
  looperWorkaround(arg2, ...args);
}
```

**Comment by austincummings** [comment]
I could see this feature providing a lot of safety. And it's not only limited to spread use cases. The way I see it is the array type would be narrowed to a tuple having the number of elements given in the expression. It could only really work given a constant number, but that'd be good enough in a lot of cases. Here's what I think the types would be after narrowing:

```ts
if (a.length > 1) {
    a; // [any, any, ...any[]]
}

if (a.length >= 1) {
    a; // [any, ...any[]]
}

if (a.length === 1) {
    a; // [any]
}

if (a.length === 0) {
    a; // []
}

if (a.length > 1) {
    a; // [any, any, ...any[]]
}

if (a.length < 0) {
    a; // never
}

if (a.length === -1) {
    a; // never
}
```

I also may be stretching or missing something here, but it might also have the consequence of allowing compile time construction of arbitrary length tuples. For example if you were to create a function like:

```ts
function makeArray<T extends number>(n: T) {
    const newArr = new Array(n);
    if (newArr.length === n) { // checker would need to use the type of `n` to get the length of the tuple type
        return newArr;
    }
    // This should be unreachable but seems out of scope of this issue
}

const tupleOf3 = makeArray(3);
```

Provided that `T` is a constant number (as opposed to `number`, not sure of the terminology here...) then I _believe_ that the type that could come out of `makeArray(3)` would be a `[any, any, any]`.

**Comment by KSXGitHub** [comment]
@bopandrade 

> Although in my example there is a simple workaround since i only use one type:
> 
> ```ts
> function recWorkaround(arg1: string, arg2?: string, ...args:string[]) {
>   console.log(arg1);
>   if (!arg2) return;
>   recWorkaround(arg2, ...args);
> }
> ```

I can call your function like this

```typescript
recWorkaround(str1, undefined, str2, str3, str4)
```

**Comment by bopandrade** [comment]
> I can call your function like this
> 
> ```ts
> recWorkaround(str1, undefined, str2, str3, str4)
> ```

While this is true it didnt really apply for my case as it was an internal function this scenario wouldnt happen unless someone purposely did it. In any case, it wouldnt be hard to assert that args should only be truthy if arg2 is truthy.

**Comment by KSXGitHub** [comment]
> it wouldnt be hard to assert that args should only be truthy if arg2 is truthy.

That would mean adding a runtime check and a runtime test.

**Comment by jcalz** [comment]
#38000 is declined as "too complex"; is this one also too complex or is there some relevant difference here?
